content/blog/2019-09-25-making-function-calls-work.md

```diff
@@ -0,0 +1,299 @@
+
+    # Print fib(10)
+    main {
+        v0: int = const 10;
+        fib10: int = call fib v0;
+        print fib10;
+    }
+
+    # Return true if n <= 1, false otherwise.
+    lte_one (n: int): bool {
+        one: int = const 1;
+        lto: bool = le n one;
+        ret lto;
+    }
+
+    # Return fib(10). Requires: n >= 0.
+    fib (n: int): int {
+        base: bool = call lte_one n;
+        br base return continue;
+    return:
+        ret n;
+    continue:
+        one: int = const 1;
+        prev: int = sub n one;
+        prev2: int = sub prev one;
+        fib1: int = call fib prev;
+        fib2: int = call fib prev2;
+        ans: int = add fib1 fib2;
+        ret ans;
+    }
+
+#### Nested function definitions
+In the above example, `lte_one` is just a helper function for `fib`and is not used anywhere else. To avoid cluttering the global function definitions, it would be nice to only define functions where they are useful.
+
+To do this, we introduced support for nested function definitions (i.e., function definitions within function definitions). First we add a new rule to the grammar. We add a new `instr` rule of the format `"def" func`, where `func` is the rule for normal function definitions. The `"def"` is there to avoid issues with labels (when the parser encounters `x:` it will not know if `x` is a label or a function with a return type). In the interpreter, we introduce the notion of a local function map. Previously, we had a global function map for the entire program. The local function map restricts access to a nested function to the immediate parent function.
+
+We also change the `call` operation handler to first search the local map before searching the global map. Note that this means functions can be shadowed. Below is the above program updated with nested function definitions.
+
+    # Print fib(10)
+    main {
+        v0: int = const 10;
+        fib10: int = call fib v0;
+        print fib10;
+    }
+
+    # Return fib(10). Requires: n >= 0.
+    fib (n: int): int {
+        def lte_one (n: int): bool {
+            one: int = const 1;
+            lto: bool = le n one;
+            ret lto;
+        }
+        base: bool = call lte_one n;
+        br base return continue;
+    return:
+        ret n;
+    continue:
+        one: int = const 1;
+        prev: int = sub n one;
+        prev2: int = sub prev one;
+        fib1: int = call fib prev;
+        fib2: int = call fib prev2;
+        ans: int = add fib1 fib2;
+        ret ans;
+    }
+
+#### Command-line arguments
+Adding support for command-line arguments is quite straightforward. We can use the `process.argv` variable to access arguments to `brili`. For each argument, we simply ensure that the argument is either an integer or a string containing "true" or "false", in which case we convert the string to the corresponding Boolean. We then pass these arguments to `main` the same way we would pass arguments to any other function.
```

Comment:
It might be nice to remind the reader here that we're talking about extending the interpreter. "Command-line arguments" don't make much sense as part of the language specification; this is just about hacking on a tool. (Right?)